source/common/config/delta_subscription_state.cc

```diff
@@ -51,13 +56,13 @@ bool DeltaSubscriptionState::subscriptionUpdatePending() const {
          !any_request_sent_yet_in_current_stream_;
 }
 
-UpdateAck
-DeltaSubscriptionState::handleResponse(const envoy::api::v2::DeltaDiscoveryResponse& message) {
+UpdateAck DeltaSubscriptionState::handleResponse(const void* response_proto_ptr) {
```

Comment:
is it possible to make it so that this is not a naked void*?

Reply:
The inevitable question; I've been expecting it :)

First, some background: The core purpose of this PR is to be able to share several hundred lines of code between delta and SotW. As a nice bonus, it also lets us wrap all of gRPC xDS (delta vs SotW, aggregated vs not) behind GrpcSubscriptionImpl, whereas there was previously GrpcSubscriptionImpl, GrpcMuxSubscriptionImpl, DeltaSubscriptionImpl, and would have needed DeltaMuxSubscriptionImpl.

The obstacle here is that we are trying to code-share two gRPC services, using two sets of request/response protos. That is not something the protocol buffers framework likes: it is not ok to try to get polymorphism involved with the generated C++ classes. Rather, you're supposed to have something like an outer container type containing a oneof of your similar types. (As an aside, I believe @htuch once mentioned that it would be good if gRPC xDS could someday have delta+SotW contained in exactly this way. If that ever happens, the void* and the delta/SotW subclassing of GrpcMux will no longer be necessary; cleaner code!)

Anyways, the polymorphism hostility expresses itself in GrpcStream and the underlying gRPC client stuff, which is inextricably *templated* on the request+response proto classes. The result: it is simply impossible to hold a gRPC client instance without having already decided at compile time which of delta/SotW you are doing.

The situation is that we have the following structure:
GrpcStream<==>GrpcMux<==>SubscriptionState
, where
1) SubscriptionState has (via polymorphism) delta/SotW-specific logic
2) GrpcMux doesn't want to care about delta vs SotW, but has to sit between GrpcStream and SubscriptionState, to accomplish the multiplexing of multiple SubscriptionStates
3) GrpcStream's template-ness forces GrpcMux to care about delta vs SotW

Crucially, due to how all of this gets set up, even though a GrpcMux instance doesn't want to be delta/SotW-specific in its logic, it *is* specific in what it actually holds: its SubscriptionStates and GrpcStream are either all delta, or all SotW. So, although GrpcMux views the protos it exchanges with SubscriptionState as "could be delta or SotW", in practice they are always the same type as GrpcStream.

That's why I see void* as the way to go here. We're passing around a thing of undetermined type, whose type will be assumed correctly by everything that deals with it. It's only due to mechanical restrictions that we have to hack together our own "here's an abstract base type, you decide what it is" technique.

Now... this could instead be done with an Any. That would be inefficient; those are real conversions where data gets copied (AFAIK). But also, I think it would be counterproductive for code clarity: Any is for when you want a proto to hold an opaque proto blob. That's not what's happening here; we know what we're dealing with, and we're keeping the void* type local to a small area of code, just to hack around a mechanical constraint.

Phew. Sorry for the wall of text; this little trick basically encompasses all of the concepts of this work. I can add (as succinctly as possible) comments wherever you think they would be helpful.

Reply:
One thing you could do is make GrpcStream extend RawAsyncStreamCallbacks instead of AsyncStreamCallbacks. This would remove the templating from GrpcStream and allow you to defer the type discrimination to the *SubscriptionState::handleResponse call.

Reply:
To clarify a bit, GrpcStream would override RawAsyncStreamCallbacks::onReceiveMessageRaw(Buffer::InstancePtr&&)

Reply:
Oh nice, looks like that was added after I finished this. I spent a good amount of time looking for something equivalent to this way back when I was detemplatizing GrpcSubscription etc.

That sounds plausible, but also like a lot of messy work. I think it is best left for later.